### PR TITLE
Network performance optimizations for node.js

### PR DESCRIPTION
**Pull request by FlashSystems**
This set of three patches adds the following features to the TiddlyWiki node.js server:

- Etag based browser side caching for all resources
- GZip or Deflate compression for all responses to GET requests

This patch-set reduces the amount of data transferred significantly (sometimes by more than 80%). On slow networks or mobile devices this can speed up loading TiddlyWiki a lot. Both features can be turned on and off separately. I've been using a TiddlyWiki with both patches enabled  without issues for over a week now.

I split this into three independent commits to allow skipping any of them if the modification is not desired. Every commit (except the translation) includes a detailed description and a little benchmark.

This is my first bigger contribution to TiddlyWiki. If I got anything wrong or my changes are unidiomatic please give me feedback.

### CODE REVIEW

**Comment by Jermolene** [comment]
Thanks @FlashSystems my apologies for the delay -- maddeningly, if I miss a few days on GitHub it often takes two or three times longer to catch up again.

This is good stuff: a good idea, and your implementation has neatly anticipated the minor concerns I would have had with the proposal. There are some coding style issues, for which I apologise: we use a relatively elderly ES5 dialect of JavaScript for compatibility with older JS engines, and we have some unusual styling choices (again for historical reasons).

File: core/modules/server/routes/get-favicon.js [hunk 1]
```diff
@@ -17,9 +17,8 @@ exports.method = "GET";
 exports.path = /^\/favicon.ico$/;
 
 exports.handler = function(request,response,state) {
-	response.writeHead(200, {"Content-Type": "image/x-icon"});
 	var buffer = state.wiki.getTiddlerText("$:/favicon.ico","");
-	response.end(buffer,"base64");
+	state.sendResponse(200, {"Content-Type": "image/x-icon"}, buffer, "base64");
```
Comment (Jermolene) [inline core/modules/server/routes/get-favicon.js:21]: Apologies, as you can see there are inconsistencies, but the intention is that we don't use spaces after commas in parameter lists

File: core/modules/server/routes/get-index.js [hunk 2]
```diff
@@ -12,38 +12,16 @@ GET /
...
-			responseHeaders["Content-Encoding"] = "deflate";
-			text = zlib.deflateSync(text);
-		} else if (/\bgzip\b/.test(acceptEncoding)) {
-			responseHeaders["Content-Encoding"] = "gzip";
-			text = zlib.gzipSync(text);
-		}
-	}
-	response.writeHead(200,responseHeaders);
-	response.end(text);
+	state.sendResponse(200, responseHeaders, text);
```
Comment (Jermolene) [inline core/modules/server/routes/get-index.js:24]: (A welcome side effect of this PR is to make the GZip handling more generic).

File: core/modules/server/server.js [hunk 3]
```diff
@@ -78,6 +82,69 @@ function Server(options) {
...
+encoding: the encoding of the data to send (passed to the end method of the response instance)
+*/
+function sendResponse(request,response,statusCode,headers,data,encoding) {
+	if (this.enableBrowserCache && (statusCode == 200)) {
+		var hash = crypto.createHash('md5');
+		// Put everything into the hash that could change and invalidate the data that
+		// the browser already stored. The headers the data and the encoding.
+		hash.update(data);
+		hash.update(JSON.stringify(headers));
+		if (encoding) hash.update(encoding);
```
Comment (Jermolene) [inline core/modules/server/server.js:104]: Two things: there shouldn't a space between the `if` and the opening parenthesis, and we always use braces around the bodies of `if` statements

**Comment by FlashSystems** [comment]
Hello @Jermolene. You have high demands regarding your response time. 9 days is quite good in my opinion. I have to apologize for not adhering to the style guide consequently and causing you even more work.

I've corrected the coding style issues and updated the pull request.

**Comment by FlashSystems** [comment]
Just wanted to ask if I can do anything to get this merged. I'd really like to see this in the next version of TiddlyWiki. It's working flawless on my TiddlyWiki 5.1.23 setup.

**Comment by FlashSystems** [comment]
Hello @Jermolene, could you please tell me what I can do to get this pull request merged? Your remarks have been fixed and I don't see any obvious cause for this pull request being stalled.

**Comment by Jermolene** [comment]
My sincere apologies for the frustration @FlashSystems, and thank you for making those changes.

**Pull request merged**